server/src/Icepeak/Server/WebsocketServer.hs

```diff
@@ -173,3 +221,67 @@ processUpdates core = go
           go
         -- Stop the loop when we receive a Nothing.
         Nothing -> pure ()
+
+-- * Timeout handling
+--
+-- The websockets library lets you send pings, but it has no built in way to
+-- terminate clients that never send a pong back. We implement this ourselves by
+-- keeping track of when we last received a pong, and then terminating the
+-- connection if the time between the last sent ping and the last received pong
+-- exceeds a certain threshold. See 'WSServerOptions' for more information
+
+-- | The connection-specific on-pong handler. This writes the time at which the
+-- last pong was received so 'pingHandler' can terminate the connection if the
+-- client stops sending pongs back.
+pongHandler :: WSServerOptions -> IO ()
+pongHandler (WSServerOptions lastPongTime) = getTime Monotonic >>= void . swapMVar lastPongTime
+
+-- | An action passed to 'withInteruptablePingThread' that is used together with
+-- 'pongHandler' to terminates a websocket connection if the client stops
+-- sending timely pongs. This returns @True@ if the connection has timed out and
+-- should be terminated.
+pingHandler :: Config -> WSServerOptions -> IO Bool
+pingHandler config (WSServerOptions lastPongTime) = do
+  now <- getTime Monotonic
+  let pingInterval     = TimeSpec (fromIntegral $ configWebSocketPingInterval config) 0
+      pongTimeout      = TimeSpec (fromIntegral $ configWebSocketPongTimeout config) 0
+      lastPongDeadline = now - pingInterval - pongTimeout
+
+  lastPong <- readMVar lastPongTime
+  return $! lastPong < lastPongDeadline
+
+-- | Similar to 'WS.withPingThread', except that it uses a combination of
+-- 'pingHandler' and the 'pongHandler' set in the websocket connection's pong
+-- handler to detect that the thread client stopped sending pongs. If that
+-- happens the @app@ action will be canceled immediately.
+--
+-- The @pingAction@ is exected on every ping, and it should return @True@ if the
+-- client has timed out and the connection should be terminated.
+withInteruptablePingThread :: WS.Connection -> Int -> IO Bool -> IO () -> IO ()
```

Comment:
Is should be `Interruptible` or `Interruptable`. The first form is the most common though, I'd go for that one.